backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/EnumClassLowering.kt

```diff
@@ -158,9 +158,88 @@ internal class EnumUsageLowering(val context: Context)
+            val eqEqCall = it.condition as IrCall
+            val entry = eqEqCall.getArguments()[1].second as? IrGetEnumValue
+            if (entry != null) {
+                val entryOrdinal = enumEntryOrdinals[entry.descriptor]!!
+                // replace condition with trivial comparison of ordinals
+                it.condition = IrCallImpl(eqEqCall.startOffset, eqEqCall.endOffset, areEqualByValue).apply {
+                    putValueArgument(0, IrConstImpl.int(entry.startOffset, entry.endOffset, context.builtIns.intType, entryOrdinal))
+                    putValueArgument(1, IrGetValueImpl(ordinalVariable.startOffset, ordinalVariable.endOffset, ordinalVariable.symbol))
+                }
+            }
+        }
+        // Process nested when constructs.
+        expression.transformChildrenVoid(this)
+        return expression
+    }
+
+    // Checks that irBlock satisfies all constrains of this lowering.
+    // 1. Block's origin is WHEN
+    // 2. Subject of `when` is variable of enum type
+    private fun shouldLower(irBlock: IrBlock): Boolean {
+        if (irBlock.origin != IrStatementOrigin.WHEN) {
+            return false
+        }
+        // when-block should have two children: temporary variable and when itself.
+        assert(irBlock.statements.size == 2)
+        val tempVariable = irBlock.statements[0] as IrVariable
+        val enumClass = tempVariable.type.constructor.declarationDescriptor as? ClassDescriptor ?: return false
+        return enumClass.kind == ClassKind.ENUM_CLASS
+    }
+
+    private fun createOrdinalVariable(irBlock: IrBlock): IrVariable {
+        val tempVariable = irBlock.statements[0] as IrVariable
+        val enumClass = tempVariable.type.constructor.declarationDescriptor as ClassDescriptor
+        val ordinalPropertyGetter = context.ir.getPropertyGetterByName(enumClass, Name.identifier("ordinal"))
+        val getOrdinal = IrCallImpl(tempVariable.startOffset, tempVariable.endOffset, ordinalPropertyGetter).apply {
+            dispatchReceiver = IrGetValueImpl(tempVariable.startOffset, tempVariable.endOffset, tempVariable.symbol)
+        }
+        // Create temporary variable for subject's ordinal.
+        val ordinalDescriptor = IrTemporaryVariableDescriptorImpl(tempVariable.descriptor,
+                Name.identifier(tempVariable.name.asString() + "_ordinal"), context.builtIns.intType)
+        return IrVariableImpl(tempVariable.startOffset, tempVariable.endOffset,
+                IrDeclarationOrigin.IR_TEMPORARY_VARIABLE, ordinalDescriptor, getOrdinal)
+    }
+}
+
+
 internal class EnumClassLowering(val context: Context) : ClassLoweringPass {
+    private val enumEntryOrdinals = mutableMapOf<ClassDescriptor, Int>()
+
     fun run(irFile: IrFile) {
         runOnFilePostfix(irFile)
+        // EnumWhenLowering should be performed before EnumUsageLowering because
+        // the latter is performing lowering of IrGetEnumValue
```

Comment:
"the latter is performing lowering of IrGetEnumValue" -> "the latter performs lowering of IrGetEnumValue."